src/tests/hash_functions/validation/README.md

```diff
@@ -0,0 +1,9 @@
+The validation directory contains code to validate the Fortran hash functions against the original C/C++ codes. It consists of three executable applications:
+
+* `generate_key_array.f90` - creates a file containing 2048 random 8 bit integers.
+
+* `generate_hash_arrays.cpp` - reads the file generated by `generate_key_array.f90` and uses its contents to generate 2049 hashes for each hash algorithm and outputa files containing the hashes.
+
+* `hash_validity_test.f90`- reads the file generated by `generate_key_array.f90` and uses its contents to generate 2049 hashes for each hash algorithm and compares the result with the corresponding outputs of `generate_hash_arrays.cpp` reporting if the outputs are not equal.
+
+Note the C code for nmhash assumes that the C compiler is either gcc or MSVC, and will not compile with the Intel C compiler.
```

Comment:
Do you know why? e.g. A compiler bug, or a non-standard feature?

Reply:
So I believe that these tests would not be run by default in the stdlib test-suite?

Is the idea that by providing them here, people can at least manually compile and test the code? And in future if we have better support for C/c++, then it could be integrated into the test suite?

This may need some discussion.

Reply:
The C code assumes the existence of library functions for a circular shift, with names particular to gcc and MSVC. With a little work a function or macro could be defined as ` I << lshift || i >> 32 - shift`, and used for other compilers, but I would prefer to leave that to someone else. Yes the idea here is to have them available so they can be manually run.

Reply:
Could the cpp preprocessor use for use the right names?

Since the code is available within stdlib (thank you for that), I think it would be good to include these tests with all other tests.

Reply:
> 

@wclodius2 In `nmhash.h` line 63, is there not a macro already defined?

Reply:
In both `nmhash.h` and `nmhash_scalar.h` lines 53-65 a macro is defined:

    #if defined(__has_builtin)
    #  if __has_builtin(__builtin_rotateleft32)
    #    define NMH_rotl32 __builtin_rotateleft32 /* clang */
    #  endif
    #endif
    #if !defined(NMH_rotl32)
    #  if defined(_MSC_VER)
         /* Note: although _rotl exists for minGW (GCC under windows), performance seems poor */
    #    define NMH_rotl32(x,r) _rotl(x,r)
    #  else
    #    define NMH_rotl32(x,r) (((x) << (r)) | ((x) >> (32 - (r))))
    #  endif
    #endif

I think either `if __has_builtin(__builtin_rotateleft32)` gives a false positive for `icc`, or `if defined(_MSC_VER)` gives a false positive for `icc`. or `_rotl(x,r)` is not 32 bit.

Reply:
As far as I can check, both `gcc` and `icc` uses `NMH_rotl32(x,r) (((x) << (r)) | ((x) >> (32 - (r))))` (I tested by saving the preprocessed files)